File: smash-ranks-android/app/src/main/java/com/garpr/android/models/WinsLosses.java
package com.garpr.android.models;

import android.os.Parcel;
import android.os.Parcelable;

import com.google.gson.annotations.SerializedName;

public class WinsLosses implements Parcelable {

    @SerializedName("losses")
    public final int mLosses;

    @SerializedName("wins")
    public final int mWins;


    public WinsLosses(final int wins, final int losses) {
        mWins = wins;
        mLosses = losses;
    }

    private WinsLosses(final Parcel source) {
        mLosses = source.readInt();
        mWins = source.readInt();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(final Parcel dest, final int flags) {
        dest.writeInt(mLosses);
        dest.writeInt(mWins);
    }

    public static final Creator<WinsLosses> CREATOR = new Creator<WinsLosses>() {
        @Override
        public WinsLosses createFromParcel(final Parcel source) {
            return new WinsLosses(source);
        }

        @Override
        public WinsLosses[] newArray(final int size) {
            return new WinsLosses[size];
        }
    };

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/models/WinsLosses.kt
package com.garpr.android.models

import android.os.Parcel
import android.os.Parcelable
import com.garpr.android.extensions.createParcel
import com.garpr.android.extensions.readAbsPlayer
import com.garpr.android.extensions.writeAbsPlayer
import com.google.gson.annotations.SerializedName

data class WinsLosses(
        @SerializedName("player") val player: AbsPlayer,
        @SerializedName("player_wins") val playerWins: Int,
        @SerializedName("opponent") val opponent: AbsPlayer,
        @SerializedName("opponent_wins") val opponentWins: Int
) : Parcelable {

    companion object {
        @JvmField
        val CREATOR = createParcel { WinsLosses(it.readAbsPlayer(), it.readInt(),
                it.readAbsPlayer(), it.readInt()) }
    }

    val winLossPercentages: FloatArray
        get() {
            return when {
                playerWins == 0 && opponentWins == 0 -> {
                    floatArrayOf(0f, 0f)
                }

                playerWins == 0 -> {
                    floatArrayOf(0f, 1f)
                }

                opponentWins == 0 -> {
                    floatArrayOf(1f, 0f)
                }

                else -> {
                    val percentPlayerWins: Float = playerWins.toFloat() /
                            (playerWins.toFloat() + opponentWins.toFloat())
                    floatArrayOf(percentPlayerWins, 1f - percentPlayerWins)
                }
            }
        }

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeAbsPlayer(player, flags)
        dest.writeInt(playerWins)
        dest.writeAbsPlayer(opponent, flags)
        dest.writeInt(opponentWins)
    }

}
